twofactor/twofactor.go

```diff
@@ -0,0 +1,359 @@
+	processId []byte,
+	pubCensusType db.CensusType,
+	orgParticipants []db.CensusMembershipParticipant,
+) error {
+	for i, participant := range orgParticipants {
+		userID := make(internal.HexBytes, hex.EncodedLen(len(participant.ParticipantNo)))
+		hex.Encode(userID, []byte(participant.ParticipantNo))
+
+		electionId := internal.HexBytes{}
+		if err := electionId.FromString(participant.ElectionID); err != nil {
+			return fmt.Errorf("wrong electionID at participant %d", i)
+		}
+		electionIDs := []internal.HexBytes{electionId}
+
+		if err := tf.stg.AddUser(userID, electionIDs, participant.HashedEmail, participant.HashedPhone, ""); err != nil {
+			log.Warnf("cannot add user from line %d", i)
+		}
+	}
+	log.Debug(tf.stg.String())
+	return nil
+}
+
+func (tf *Twofactor) InitiateAuth(
+	electionID []byte,
+	userId []byte,
+	contact string,
+	notifType notifications.NotificationType,
+) AuthResponse {
+	// If first step, build new challenge
+	if len(userId) == 0 {
+		return AuthResponse{Error: "incorrect auth data fields"}
+	}
+	// var userID internal.HexBytes
+	// if err := userID (userId); err != nil {
+	// 	return AuthResponse{Error: "incorrect format for userId"}
+	// }
+	userID := internal.HexBytes(userId)
+
+	// Generate challenge and authentication token
+	// We need to ensure the challenge secret is a valid base32-encoded string
+	// Instead of concatenating userID.String() (hex) with otpSalt (base32),
+	// we'll use a different approach to create a unique secret per user
+	challengeSecret := gotp.RandomSecret(16) // Use a fresh random secret
+	atoken := uuid.New()
+
+	// Get the phone number. This methods checks for electionID and user verification status.
+	_, _, attemptNo, err := tf.stg.NewAttempt(userID, electionID, challengeSecret, &atoken)
+	if err != nil {
+		log.Warnf("new attempt for user %s failed: %v", userID, err)
+		return AuthResponse{Error: err.Error()}
+	}
+	if contact == "" {
+		log.Warnf("phone is nil for user %s", userID)
+		return AuthResponse{Error: "no phone for this user data"}
+	}
+	// Enqueue to send the challenge
+	challenge := gotp.NewDefaultHOTP(challengeSecret)
+	// Generate the OTP code using the attempt number
+	otpCode := challenge.At(attemptNo)
+
+	if notifType == notifications.Email {
+		if err := tf.mailQueue.add(userID, electionID, contact, otpCode); err != nil {
+			log.Errorf("cannot enqueue challenge: %v", err)
+			return AuthResponse{Error: "problem with Email challenge system"}
+		}
+		log.Infof("user %s challenged with %s at contact %s", userID.String(), otpCode, contact)
+	} else if notifType == notifications.SMS {
+		if err := tf.smsQueue.add(userID, electionID, contact, otpCode); err != nil {
+			log.Errorf("cannot enqueue challenge: %v", err)
+			return AuthResponse{Error: "problem with SMS challenge system"}
+		}
+		log.Infof("user %s challenged with %s at contact %s", userID.String(), otpCode, contact)
+	} else {
+		return AuthResponse{Error: "invalid notification type"}
+	}
```

Comment:
With my approach you just need to call `NewNotificationChallege` with the type and the data and then enqueue it.